src/main/java/dev/morling/onebrc/CalculateAverage_godofwharf.java

```diff
@@ -0,0 +1,632 @@
+                    }
+                    else {
+                        return decimal + fractional;
+                    }
+                }
+                else {
+                    int decimal = fastMul10(toDigit(ch1)) + toDigit(ch2);
+                    double fractional = DOUBLES[toDigit(ch4)];
+                    return Math.negateExact(decimal) - fractional;
+                }
+            }
+            catch (ArrayIndexOutOfBoundsException e) {
+                printDebugMessage("Array index out of bounds for string: %s%n", new String(b, 0, len));
+                throw new RuntimeException(e);
+            }
+            catch (StringIndexOutOfBoundsException e) {
+                printDebugMessage("String index out of bounds for string: %s%n", new String(b, 0, len));
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    // record classes
+    record Measurement(byte[] station,
+                       double temperature,
+                       int hash,
+                       State.AggregationKey aggregationKey) {
+
+    public Measurement(byte[] station,
+                       double temperature,
+                       int hashCode) {
+            this(station,
+                    temperature,
+                    hashCode,
+                    new State.AggregationKey(station, hashCode));
+        }
+
+    }
+
+    record LineMetadata(byte[] station,
+                        byte[] temperature,
+                        int stationLen,
+                        int temperatureLen,
+                        int offset,
+                        int precomputedHashCode, boolean isAscii) {
+    }
+
+    record Split(long offset, long length, List<Page> pages) {
+    }
+
+    record Page(long offset, long length) {
+    }
+
+    public static class SearchResult {
+        private int[] offsets;
+        private int len;
+
+        public SearchResult(final int[] offsets,
+                            final int len) {
+            this.offsets = offsets;
+            this.len = len;
+        }
+    }
+
+    public static class FastHashMap {
+        private TableEntry[] tableEntries;
+        private int size;
+
+        public FastHashMap(final int capacity) {
+            this.size = tableSizeFor(capacity);
+            this.tableEntries = new TableEntry[size];
+        }
+
+        private int tableSizeFor(final int capacity) {
+            int n = -1 >>> Integer.numberOfLeadingZeros(capacity - 1);
+            return (n < 0) ? 1 : (n >= Integer.MAX_VALUE) ? Integer.MAX_VALUE : n + 1;
+        }
+
+        public void put(final State.AggregationKey key,
+                        final MeasurementAggregator aggregator) {
+            int h1 = key.hashCode;
+            int hash = h1 ^ (h1 >>> 12);
+            tableEntries[(size - 1) & hash] = new TableEntry(key, aggregator);
+        }
+
+        public MeasurementAggregator get(final State.AggregationKey key) {
```

Comment:
How do you handle hash collisions here? Looks like you may end up aggregating the values for several stations with the same hash?

Reply:
Hey @gunnarmorling, thanks for the review comment. I forgot to remove the FastHashMap implementation. It is not getting used because it results in collisions. I tried implementing double hashing for probing but the implementation proved to be slower than using Java HashMap. So I thought I would fix it in a future iteration. For now, I will remove this class because it is not getting used anywhere.

Reply:
Ah got it. Thx for clarifying.